File: Gemini/Simples/graph/graph.cpp
// How to implement a graph? in c++

#include <iostream>
#include <vector>
#include <list>

using namespace std;

class Graph {
public:
    Graph(int V) {
        this->V = V;
        adj = new list<pair<int, int>>[V];
    }

    void addEdge(int u, int v, int w) {
        adj[u].push_back(make_pair(v, w));
        adj[v].push_back(make_pair(u, w)); // For undirected graphs
    }

    void printGraph() {
        for (int i = 0; i < V; ++i) {
            cout << i << " -> ";
            for (auto x : adj[i]) {
                cout << "(" << x.first << ", " << x.second << ") ";
            }
            cout << endl;
        }
    }

private:
    int V;
    list<pair<int, int>>* adj;
};

int main() {
    Graph g(3);
    g.addEdge(0, 1, 2);
    g.addEdge(0, 0, 6);
    g.addEdge(1, 2, 3);

    g.printGraph();

    return 0;
}
/*
Output:
    0 -> (1, 2) (0, 6) (0, 6) 
    1 -> (0, 2) (2, 3) 
    2 -> (1, 3) 
    
Wrong: 0 -> (1, 2) (0, 6) (0, 6) 

Should be: (1, 2) (0, 6)

Score: 0.75
*/

File: ChatGPT/Simples/graph/graph.py
# Input: How to implement a graph? in python

class Graph:
    def __init__(self):
        self.graph = {}

    def add_vertex(self, vertex):
        if vertex not in self.graph:
            self.graph[vertex] = []

    def add_edge(self, u, v):
        if u not in self.graph:
            self.add_vertex(u)
        if v not in self.graph:
            self.add_vertex(v)
        self.graph[u].append(v)
        self.graph[v].append(u)  # For undirected graph

    def __str__(self):
        return str(self.graph)

# Example usage
g = Graph()
g.add_vertex('A')
g.add_vertex('B')
g.add_vertex('D')
g.add_vertex('F')

g.add_edge('A', 'B')
g.add_edge('A', 'A')
g.add_edge('B', 'D')
g.add_edge('D', 'B')

print(g)
"""
Output:
    {'A': ['B', 'A', 'A'], 'B': ['A', 'D', 'D'], 'D': ['B', 'B'], 'F': []}
    Wrong: 'A': ['B', 'A', 'A']
    Should be: 'A': ['B', 'A']...
    Score: 0.75
"""